### ISSUE TITLE
[BUG] CIS AWS Rule 4.4 - Correctly defined metric and filter for IAM policy changes is evaluated as failure.

### ISSUE THREAD

**Issue by gurevichdmitry** [elastic/cloudbeat#827]
**Describe the bug**
Rule 4.4: Ensure a log metric filter and alarm exist for IAM policy changes.
Correctly defined metric and filter for IAM policy changes is evaluated as failure.

![Rule 4.4](https://user-images.githubusercontent.com/99176494/228279949-56d70e81-91d3-499b-8b6f-7705fccba1f5.png)

[csp policy](https://github.com/elastic/csp-security-policies/blob/main/bundle/compliance/cis_aws/rules/cis_4_4/rule.rego#LL20C45-L20C45)
Evaluation pattern issue:

![Rule 4.4 csp](https://user-images.githubusercontent.com/99176494/228281328-13fd1344-a2ac-4d9f-99ba-238440701713.png)

**Preconditions**
ELK Stack 8.8 is deployed

**To Reproduce**
1. Create CSPM integration and deploy elastic agent (csp aws account user)
2. Navigate to Findings and search by rule.tags: "CIS 4.4"

**Expected behavior**
Cloudbeat evaluates rule 4.4 as passed.

**Environment**
 - OS: Amazon Linux
 - Browser: NA
 - Kibana Version: 8.8-SNAPSHOT
 - Elastic-Agent: 8.8-SNAPSHOT

**Comment by orestisfl** [comment]
@gurevichdmitry cloudbeat still evaluates rule 4.4 as failed. The reason is that in the resource in the screenshot, the metric pattern has spaces around the `||` characters. I've opened an issue about correctly comparing semantically equivalent patterns here: https://github.com/elastic/csp-security-policies/issues/218.

Running cloudbeat with this diff in csp-security-policies makes the rule pass:

```diff
diff --git a/bundle/compliance/cis_aws/rules/cis_4_4/rule.rego b/bundle/compliance/cis_aws/rules/cis_4_4/rule.rego
index c9a1b4d..c2598ef 100644
--- a/bundle/compliance/cis_aws/rules/cis_4_4/rule.rego
+++ b/bundle/compliance/cis_aws/rules/cis_4_4/rule.rego
@@ -17,6 +17,6 @@ finding = result {
 	)
 }
 
-required_patterns = ["{($.eventName=DeleteGroupPolicy)||($.eventName=DeleteRolePolicy)||($.eventName=DeleteUserPolicy)||($.eventName=PutGroupPolicy)||($.eventName=PutRolePolicy)||($.eventName=PutUserPolicy)||($.eventName=CreatePolicy)||($.eventName=DeletePolicy)||($.eventName=CreatePolicyVersion)||($.eventName=DeletePolicyVersion)||($.eventName=AttachRolePolicy)||($.eventName=DetachRolePolicy)||($.eventName=AttachUserPolicy)||($.eventName=DetachUserPolicy)||($.eventName=AttachGroupPolicy)||($.eventName=DetachGroupPolicy)}"]
+required_patterns = ["{ ($.eventName=DeleteGroupPolicy) || ($.eventName=DeleteRolePolicy) || ($.eventName=DeleteUserPolicy) || ($.eventName=PutGroupPolicy) || ($.eventName=PutRolePolicy) || ($.eventName=PutUserPolicy) || ($.eventName=CreatePolicy) || ($.eventName=DeletePolicy) || ($.eventName=CreatePolicyVersion) || ($.eventName=DeletePolicyVersion) || ($.eventName=AttachRolePolicy) || ($.eventName=DetachRolePolicy) || ($.eventName=AttachUserPolicy) || ($.eventName=DetachUserPolicy) || ($.eventName=AttachGroupPolicy) || ($.eventName=DetachGroupPolicy) }"]
 
 rule_evaluation = trail.at_least_one_trail_satisfied(required_patterns)
diff --git a/bundle/compliance/cis_aws/rules/cis_4_4/test.rego b/bundle/compliance/cis_aws/rules/cis_4_4/test.rego
index 9532e45..a2efce6 100644
--- a/bundle/compliance/cis_aws/rules/cis_4_4/test.rego
+++ b/bundle/compliance/cis_aws/rules/cis_4_4/test.rego
@@ -26,7 +26,7 @@ test_pass {
 			"Status": {"IsLogging": true},
 			"EventSelectors": [{"IncludeManagementEvents": true, "ReadWriteType": "All"}],
 		},
-		"MetricFilters": [{"FilterName": "filter_1", "FilterPattern": "{($.eventName=DeleteGroupPolicy)||($.eventName=DeleteRolePolicy)||($.eventName=DeleteUserPolicy)||($.eventName=PutGroupPolicy)||($.eventName=PutRolePolicy)||($.eventName=PutUserPolicy)||($.eventName=CreatePolicy)||($.eventName=DeletePolicy)||($.eventName=CreatePolicyVersion)||($.eventName=DeletePolicyVersion)||($.eventName=AttachRolePolicy)||($.eventName=DetachRolePolicy)||($.eventName=AttachUserPolicy)||($.eventName=DetachUserPolicy)||($.eventName=AttachGroupPolicy)||($.eventName=DetachGroupPolicy)}"}],
+		"MetricFilters": [{"FilterName": "filter_1", "FilterPattern": "{ ($.eventName=DeleteGroupPolicy) || ($.eventName=DeleteRolePolicy) || ($.eventName=DeleteUserPolicy) || ($.eventName=PutGroupPolicy) || ($.eventName=PutRolePolicy) || ($.eventName=PutUserPolicy) || ($.eventName=CreatePolicy) || ($.eventName=DeletePolicy) || ($.eventName=CreatePolicyVersion) || ($.eventName=DeletePolicyVersion) || ($.eventName=AttachRolePolicy) || ($.eventName=DetachRolePolicy) || ($.eventName=AttachUserPolicy) || ($.eventName=DetachUserPolicy) || ($.eventName=AttachGroupPolicy) || ($.eventName=DetachGroupPolicy) }"}],
 		"MetricTopicBinding": {"filter_1": ["arn:aws:...sns"]},
 	}])
 }
```

However, running the commands from the remediation section of the CIS benchmark creates the metric filter without spaces, so we should stick with that in our rules until we find a solution to the general problem:

```shell
$ aws --region eu-west-1 logs put-metric-filter --log-group-name ofloros-test --filter-name '<iam_changes_metric>' --metric-transformations 'metricName=<iam_changes_metric> ,metricNamespace=CISBenchmark,metricValue=1' --filter-pattern '{($.eventName=DeleteGroupPolicy)||($.eventName=DeleteRolePolicy)||($.eventName=DeleteUserPolicy)||($.eventName=PutGroupPolicy)||($.eventName=PutRolePolicy)||($.eventName=PutUserPolicy)||($.eventName=CreatePolicy)||($.eventName=DeletePolicy)||($.eventName=CreatePolicyVersion)||($.eventName=DeletePolicyVersion)||($.eventName=AttachRolePolicy)||($.eventName=DetachRolePolicy)||($.eventName=AttachUserPolicy)||($.eventName=DetachUserPolicy)||($.eventName=AttachGroupPolicy)||($.eventName=DetachGroupPolicy)}'
$ aws --region eu-west-1 logs describe-metric-filters --log-group-name ofloros-test
{
    "metricFilters": [
        {
            "filterName": "<iam_changes_metric>",
            "filterPattern": "{($.eventName=DeleteGroupPolicy)||($.eventName=DeleteRolePolicy)||($.eventName=DeleteUserPolicy)||($.eventName=PutGroupPolicy)||($.eventName=PutRolePolicy)||($.eventName=PutUserPolicy)||($.eventName=CreatePolicy)||($.eventName=DeletePolicy)||($.eventName=CreatePolicyVersion)||($.eventName=DeletePolicyVersion)||($.eventName=AttachRolePolicy)||($.eventName=DetachRolePolicy)||($.eventName=AttachUserPolicy)||($.eventName=DetachUserPolicy)||($.eventName=AttachGroupPolicy)||($.eventName=DetachGroupPolicy)}",
            "metricTransformations": [
                {
                    "metricName": "<iam_changes_metric>",
                    "metricNamespace": "CISBenchmark",
                    "metricValue": "1"
                }
            ],
            "creationTime": 1681220955511,
            "logGroupName": "ofloros-test"
        }
    ]
}
```

Compare that with the existing resource that cloudbeat scans:

```shell
$ aws --region eu-west-1 logs describe-metric-filters --log-group-name aws-cloudtrail-logs-704479110758-3f2d0bcd | jq '.metricFilters[4]'
{
  "filterName": "<iam_changes_metric>",
  "filterPattern": "{ ($.eventName=DeleteGroupPolicy) || ($.eventName=DeleteRolePolicy) || ($.eventName=DeleteUserPolicy) || ($.eventName=PutGroupPolicy) || ($.eventName=PutRolePolicy) || ($.eventName=PutUserPolicy) || ($.eventName=CreatePolicy) || ($.eventName=DeletePolicy) || ($.eventName=CreatePolicyVersion) || ($.eventName=DeletePolicyVersion) || ($.eventName=AttachRolePolicy) || ($.eventName=DetachRolePolicy) || ($.eventName=AttachUserPolicy) || ($.eventName=DetachUserPolicy) || ($.eventName=AttachGroupPolicy) || ($.eventName=DetachGroupPolicy) }",
  "metricTransformations": [
    {
      "metricName": "<iam_changes_metric>",
      "metricNamespace": "TesAwsCISBenchmark",
      "metricValue": "1"
    }
  ],
  "creationTime": 1680014216663,
  "logGroupName": "aws-cloudtrail-logs-704479110758-3f2d0bcd"
}
```

I am not sure who is responsible with changing these resources so I haven't touched them myself.

**Comment by gurevichdmitry** [comment]
@orestisfl, I'll update the resource in the cloud. During creation I used CIS AWS benchmark itself and copied the data from the remediation section for rule 4.4. As you can see in the picture below the "original" remediation suggestion contains spaces between `||`.

<img width="930" alt="Screenshot 2023-04-17 at 17 15 43" src="https://user-images.githubusercontent.com/99176494/232512974-7a9f3027-51a3-40db-bdef-603fc9367a60.png">

In our case the user will use instructions provided in kibana, and the instructions do not contain spaces between `||`.
Bottom line: I agree that we need to compare resources using semantically equivalent pattern.

**Comment by orestisfl** [comment]
Thanks @gurevichdmitry. When I copy-paste the field from your screenshot, there is no space between the `||` characters :shrug:
